nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/blob/DocumentBlobManagerComponent.java

```diff
@@ -434,13 +450,20 @@ public BinaryManagerStatus garbageCollectBinaries(boolean delete) {
 
     @Override
     public void markReferencedBinary(String key, String repositoryName) {
-        BlobProvider blobProvider = getBlobProvider(key, repositoryName);
-        BinaryGarbageCollector gc = blobProvider.getBinaryGarbageCollector();
-        if (gc != null) {
-            key = stripBlobKeyPrefix(key);
-            gc.mark(key);
+        List<BinaryGarbageCollector> gcs = getGarbageCollectors();
+        long idCount = gcs.stream().map(BinaryGarbageCollector::getId).distinct().count();
```

Comment:
I think doing this for every call to `markReferencedBinary` is a bit too much, in a large repo there will be millions. It's better to move the computation of the `sharedStorage` boolean and the `gcs` list to the `garbageCollectBinaries` method, which is the one that starts the whole process.